Adding design decision docs initially covering the scope of GatewayClass

robscott opened the pull request:
As I was thinking more about the discussion around #136, it seemed like it would be good to actually start adding the rationale for key decisions like this to the docs. I've gone back and forth on this one and do think there are compelling reasons to consider a namespaced GatewayClass, but ultimately ended up recommending a cluster-scoped GatewayClass here as a starting point. 

I'm very open to adding/updating advantages in favor of a namespace or cluster-scoped GatewayClass, or even changing the decision as it is proposed here. Just wanted this to be a place to continue the discussion around this decision, and when we have a final decision, have the rationale clearly recorded in the docs.

Fixes https://github.com/kubernetes-sigs/service-apis/issues/136.

/kind documentation
/cc @akutz @danehans @jpeach @youngnick

akutz commented:
Hi @robscott ,

Thank you for working on this! One thing I think worth calling out is that this is the first cluster-scoped `Class` resource I've seen proposed since the advent of Cluster API. We're now living in a world with an operator pattern for deploying/managing Kubernetes clusters and required infrastructure. 

I wonder if SIG Storage would have done things differently had they been designing things with the knowledge that their APIs would also live on "management clusters" used to deploy and manage other clusters. 

Or perhaps another API is needed entirely? A `StorageManager`? A `GatewayManager`? Something to manage the infrastructure? I agree that ideally you have a handful of `GatewayClass` resources per cluster. But when you consider the "management cluster" scenario, you are going to have many, many more. I am fine with something being distinct from `GatewayClass`, but I *would* like it to also be part of this API. 

Thank you again!

akutz commented on docs-src/design-decisions.md:
```diff
@@ -0,0 +1,42 @@
+
+### Advantages of a Namespace-Scoped GatewayClass
+* Less potential for naming conflicts among different implementations.
+* Better modeling for implementations with multiple GatewayClass resources
```
Hi @robscott,

Another way to state this, and perhaps it is a separate item for the list, is that this change enables using Service API on a management cluster designed to provision and manage other clusters and infrastructure needed by other clusters.

robscott replied:
👍Good catch, thanks! Will update that.

thockin replied:
I don't think that "management cluster" is a well-defined concept that has consistent meaning or semantics, so I kind of object to it as a Noun.  I'd rather spell out what that means.

akutz commented on docs-src/design-decisions.md:
```diff
@@ -0,0 +1,42 @@
+
+### Advantages of a Cluster-Scoped GatewayClass
+* Consistent with StorageClass and IngressClass resources.
+* Discussions with sig-storage showed that StorageClass had worked well as a
```
Did that discussion bring up the topic of managing other clusters with the operator pattern enabled by Cluster API?

robscott replied:
It did not, maybe @saad-ali has some insight here on if that would change anything from the perspective of sig-storage?

thockin replied:
You'll have to expound on the specifics of what "managing" means to you.

akutz commented on docs-src/design-decisions.md:
```diff
@@ -0,0 +1,42 @@
+* Discussions with sig-storage showed that StorageClass had worked well as a
+  cluster-scoped resource and that they were continuing to follow that pattern
+  for new Class resources.
+* Follows general pattern that resources managed by infrastructure provider
```
This pattern does not, to my knowledge, account for the concept of a management cluster.

robscott replied:
I think you're right. Is there any updated guidance here we can apply? Since this pattern has already been applied so widely in k/k, has anyone found a good way to work around or solve these potential issues with management clusters?

akutz replied:
Unfortunately I don’t. It’s why I’ve been tagging colleagues at VMware and sharing this issue. I hate that this is the proving ground for this, but this is a really interesting and broad topic on the class pattern in the world of management clusters. 

I already pinged Vince Priganano and the rest of the cluster API team working at VMware to take a look. I know you pinged Saad. I’m really interested to see what people think.

robscott commented:
@akutz Thanks for the great feedback here! It is a great point of distinction that this is the first Class resource being proposed with the context of Cluster API. @saad-ali has StorageClass been difficult to fit into this newer model of "management clusters"? Would you do anything differently with this context if you were starting today?

akutz commented on docs-src/design-decisions.md:
```diff
@@ -0,0 +1,42 @@
+  namespaces only or restrict visibility of GatewayClasses in some namespaces).
+
+### Advantages of a Cluster-Scoped GatewayClass
+* Consistent with StorageClass and IngressClass resources.
```
Did these designs account for multi-tenancy where tenants may have access to their own infrastructure?

robscott replied:
I can't speak for StorageClass, but I don't believe this was a consideration for IngressClass.

aledbf replied:
Same question here (ingress-nginx). Most of the users don't have enough permissions to create cluster-scoped resources and are limited to their own namespace.

thockin replied:
A "tenant" is undefined.  Before I will engage in any discussion of tenancy, you must tell me what that word means to you.

akutz commented on docs-src/design-decisions.md:
```diff
@@ -0,0 +1,42 @@
+* Impossible for namespace level admins to grant RBAC permissions for
+  GatewayClass resources. (With a namespace-scoped resource, anyone that could
+  configure RBAC within a namespace could grant GatewayClass access).
+* Clear that GatewayClasses can be used by multiple namespaces.
```
Well, not really. Not with the current implementation that defines the `AllowedGatewayNamespaces` and `AllowedRouteNamespaces` on the `GatewayClass`.

robscott replied:
Yeah both a namespace-scoped and cluster-scoped GatewayClass would still be referenced by Gateways in other namespaces, it just would seem at least slightly less obvious that a namespace-scoped resource could be referenced by Gateways in other namespaces. This would go against patterns elsewhere in Kubernetes. For example, with RBAC, a Role can only be referenced by a RoleBinding in the same namespace. Similar models for Pods/Deployments/DaemonSets/etc that are inherently limited to a single namespace.

akutz replied:
I mean, if you keep `GatewayClass` as a cluster scoped resource then IMO the `Allowed` fields should be removed. If you want that, then make GWC a namespace scoped resource or create  new API that is.

robscott replied:
Unfortunately even a namespace-scoped resource wouldn't solve this. We definitely want it to be possible to reference a GatewayClass from multiple namespaces. A namespace-scoped resource would make it straightforward to allow only-local references of a GatewayClass, but I don't think it helps with the very real use case of "only Gateways in namespaces labeled `env: prod` should be able to use this GatewayClass".

akutz replied:
That’s fair, but it’s one thing to define that on a namespaced type and another on a cluster scoped resource that is supposed to satisfy all comers.

thockin replied:
Absent a general form of hierarchy, we don't have a way to express who can see a thing other than "don't cross namespaces.

There are a LOT of things I would change about how k8s does this stuff, if I could do so with a reasonable amount of energy.  TTBOMK "who can use" has not come up for StorageClass or RuntimeClass yet.

akutz commented:
I also think there is a broader discussion to be had about whether people should use cluster scoped classes at all. I'd argue that `GatewayClass` isn't truly cluster scoped since the class allows the definition of namespaces for resources allowed to use the `GatewayClass`.

But since we're talking about storage, I want to use a concrete example. Imagine I have a management cluster, and I want:

* to provision an NFS export for each CAPI `Cluster` resource
* seed that NFS export/volume with information/data for use by the target cluster
* remove the NFS export when the CAPI `Cluster` resource is deleted

I've just described two things:

1. infrastructure lifecycle management 
2. leveraging that infrastructure

And I've tied the lifecycle of both to a namespaced resource, the `Cluster` object on the management cluster. For example, before I create the `Cluster` with CAPI, I would need to:

* Provision an NFS server *or* provision an export on an existing NFS server
* Mount the export onto the management cluster and write the seed data

To do that today I might:

1. Create a `PersistentVolume` resource on the management cluster:

    ```yaml
    apiVersion: v1
    kind: PersistentVolume
    metadata:
    name: my-nfs-share
    spec:
    capacity:
        storage: 5Gi # This size is used to match a volume to a tenents claim
    accessModes:
        - ReadWriteOnce # Access modes are defined below
    persistentVolumeReclaimPolicy: Recycle # Reclaim policies are defined below 
    nfs:
        server: nfs.example.com # Please change this to your NFS server
        path: /share1 # Please change this to the relevant share
    ```

2. Reconcile that persistent volume to:
    a. Provision infrastructure
    b. Mount the volume on the management cluster

    Now, I could create a new, namespaced API called `NFSServer` to handle `2a`, but if I do, I end up with the following dependency model on the management cluster:

    ```
    PersistentVolume -> NFSServer
    ```

So for every `Cluster` I provision I have to:

1. Create a namespaced `NFSServer` resource to sit alongside the `Cluster` resource
2. Create a cluster scoped `PersistentVolume` resource on the management cluster to use the new infrastructure in order to seed the required data for the target cluster

My overall point is that the cluster-scoped `Class` pattern either does not support management clusters or needs to evolve to do so. As it stands, it's clear that Cluster API enabled management clusters cannot rely on APIs that embrace the class pattern to manage and use infrastructure from the management cluster as it results in pollution of the cluster scope with hundreds, if not thousands, of resources that are aligned with CAPI `Cluster` objects.

jpeach commented on docs-src/design-decisions.md:
```diff
@@ -0,0 +1,42 @@
+* Clear that GatewayClasses can be used by multiple namespaces.
+* Simple globally unique names.
+* Works well for majority of use cases - likely only a few total GatewayClasses
+  per cluster.
```
Prior experience in the Ingress API (which effectively has a cluster scoped class) shows this works well.

akutz replied:
Neither of which account for management clusters (see my other comments). Cluster resources are anti-pattern IMO, and the `GatewayClass` isn't even a pure cluster resource as it defines relationships to namespaces as part of its spec.

akutz replied:
> Prior experience in the Ingress API (which effectively has a cluster scoped class) shows this works well.

Another thought, it may work well for Ingress, but maybe not other use cases that we're trying to present to this group and satisfy with the Service API? And that's okay, but what I am hearing is that y'all do not think our use is in-scope. If that's the decision, and it seems to be, that is fair and allows us to move on and think about how else to tackle the problem.

What I do not think is fair, is to argue that one prior experience should exclude consideration of other experiences for what is in and out of scope.

Thanks again for working with us on this!

jpeach replied:
The service-apis WG was established in order to evolve the Ingress API, so I'd argue that experience there is relevant. That's orthogonal to whether the WG scope changes over time (and I don't think this document prevents that).

akutz replied:
Fair

akutz replied:
But I’d also argue that evolve doesn’t just mean “satisfy in a different way.” I was under the impression that the scope also means to satisfy non-Ingress use cases. And even from an Ingress POV I think it’s important to consider the use cases I am raising.

akutz commented:
Just a nit, but shouldn't this start out as an issue instead of a PR? Maybe not worth making the change for this, but in the future such a large issue probably warrants discussion before going straight to a PR?

akutz commented:
I guess even if we say "Use something else to manage the infrastructure," I see a world where we have to create hundreds, even thousands, of cluster-scoped classes to access that infrastructure from a management cluster. That just seems wrong.

zjs commented:
bq. The path of least resistance is definitely having a namespaced GatewayClass.

To me, I think this is essentially a question of user personas.

Unless GatewayClass is a resource that should span tenancy boundaries, it seems like GatewayClass needs to be namespaced.

If the sort of user who defines a new GatewayClass may be the same sort of user who consumes the GatewayClass, it seems like GatewayClass needs to be namespaced.

If the sort of user who defines a new GatewayClass is always expected to be more like an infrastructure administrator, who is defining a consistent set of GatewayClasses for everyone else to consume, maybe a non-namespaced resource makes sense.

akutz commented:
I think it’s important to point out that different use cases might result in different persona expectations. What’s true for one use case may not be true for another.

robscott commented:
Great comments all around, still trying to catch up, but I do have a quick follow up for this comment from @zjs:

> If the sort of user who defines a new GatewayClass is always expected to be more like an infrastructure administrator, who is defining a consistent set of GatewayClasses for everyone else to consume, maybe a non-namespaced resource makes sense.

The current security model states that GatewayClass would be managed exclusively by Infrastructure providers or Cluster Operators: https://github.com/kubernetes-sigs/service-apis/blob/master/docs-src/security-model.md#write-permissions-for-simple-3-tier-model.

akutz commented:
Hi @robscott,

> The current security model states that GatewayClass would be managed exclusively by Infrastructure providers

The [link you provided](https://github.com/kubernetes-sigs/service-apis/blob/master/docs-src/security-model.md#write-permissions-for-simple-3-tier-model) states that an infrastructure provider is:

> responsible for the overall environment that the cluster(s) are operating in. Examples include: the cloud provider (AWS, Azure, GCP, ...), the PaaS provider in a company.

When a management cluster is responsible for _provisioning_ **and** using that infrastructure, an implicit model dependency is then created between the cluster scoped `GatewayClass` and whatever resource is used to control the lifecycle of the infrastructure to which the `GatewayClass` is attached.

Doesn't it make sense for the `GatwayClass` to also then have the ability to be the resource a provider uses to control the lifecycle of the infrastructure?

youngnick commented:
I think it's important to mention a few things about the *original* conception of the service APIs.

Firstly, the GatewayClass was intended to basically be "a definition of a Controller that watches Gateways". That's why it was cluster-scoped, because it's much, much easier to write a controller that watches everything in a cluster than it is to have a controller that watches only a subset of namespaces. The intent of having the `AllowedNamespaces` fields is to configure the *controller* that watches the Kube API to throw away objects that don't match the allowed namespaces. Note that I'm not making representations about which idea is better here, just pointing out that this is what the _current design expects_.

Secondly, Gateways and the rest of the below structure was intended to be an abstract concept of *some of the properties* of a load balancer, not a complete description of everything. It matches up well to the use case @akutz is talking about because of this.

I think that there is a difference between "Configuration for the controller that watches the Kube API" and "Configuration for the load balancer that is configured by that controller". `GatewayClass` _as it stands_ is the former, and feels like @akutz really needs the latter.

If I'm correct about that, the way to go here is to make it clearer that GatewayClass is for the controller that watches Gateways, and that we need something else to describe the infrastructure. I totally acknowledge that we have left that out until now, I think in both the in-cluster ingress controller and my-cloud-provider-configured-a-load-balancer-for-me cases, it was easy to handwave the infrastructure management away. In more bare-metal use cases, you just can't do that.

maplain commented:
> this is the first cluster-scoped Class resource I've seen proposed since the advent of Cluster API.

One unique situation created by CAPI is multiple k8s Workload clusters from different tenants are managed by one Management cluster, yet aiming to share the same infrastructure.

Before CAPI, k8s clusters form a natural tenancy boundary. 
For example, if you take a look at all the existing CSI providers, they either take Infra credentials directly from the StorageClass or from a ConfigMap with predefined name/namespace. That works fine because different k8s clusters deploy different CSIs with their own StorageClass/ConfigMap.

However, with the advent of CAPI, different k8s clusters appear in the same k8s management cluster as objects in different namespaces, which brings us back to the old namespace tenancy model.

It'll still work if we assume CAPI Workload cluster manages their own infrastructure independently. But this is no longer entirely true.

Even though CAPI Workload clusters are free to go with their own Infrastructure choice, they're assumed to **share** the same Infrastructure **by default**.

For this use case, the question is how are we supposed to consolidate the Infrastructure configurations across CAPI Workload Cluster while preserving multi-tenancy at the same time, particularly from a Management Cluster's perspective.

A namespaced GatewayClass would definitely help a lot!

It's optional so it adds on top of the original model. More importantly, it nicely and naturally fits into the namespace tenancy model.

akutz commented:
Hi @youngnick,

> That's why it was cluster-scoped, because it's much, much easier to write a controller that watches everything in a cluster than it is to have a controller that watches only a subset of namespaces.

Sure, but I don't see how these are related.

> The intent of having the AllowedNamespaces fields is to configure the controller that watches the Kube API to throw away objects that don't match the allowed namespaces. 

Okay, but this is anti-pattern to the cluster-scoped class pattern. If the intent is to restrict access, then the more natural way to do that is to introduce a namespaced version of the resource. And if you want to have it shared by multiple namespaces the place a copy of it in each namespace. Or keep the field and put it in a reserved namespace. I don't know, but at the cluster scope connotes that it should be available to the cluster, not a subset of the cluster.

> Secondly, Gateways and the rest of the below structure was intended to be an abstract concept of some of the properties of a load balancer, not a complete description of everything. 

Except every time I've tried to reconcile this abstraction into a use case I am told "No no, we meant abstract **this** way, not your way." Well, which is it? Abstract or not? Because all our proposals are met with opposition that it doesn't match the stated intent, when all we're doing is to try to make things abstract enough to work for our use case as well.

> I think that there is a difference between "Configuration for the controller that watches the Kube API" and "Configuration for the load balancer that is configured by that controller". 

I think it's more accurate to state that it's the difference between provisioning infrastructure and consuming the provisioned infrastructure. When the infrastructure is large in quantity, then having cluster scoped resources to consume it is cumbersome. And the Cluster API use case of a management cluster is a perfect breeding ground for this use case.

> If I'm correct about that, the way to go here is to make it clearer that GatewayClass is for the controller that watches Gateways, and that we need something else to describe the infrastructure. 

Again, when infrastructure scales with the number of `Cluster` resources provisioned, consuming that infrastructure on a single cluster via cluster-scoped resources like `GatewayClass` is a really poor situation.

But this is true for anything that implements the class pattern in the face of the operator model for infrastructure enabled by Cluster API.

thockin commented:
There's a lot here, so I can't pick line-by-line at comments and specific responses.  One thing really jumps out at me though:

> I see a world where we have to create hundreds, even thousands, of cluster-scoped classes to access that infrastructure from a management cluster.

I am somewhat confused why you need classes proportional to clusters under management?  The class is the cookie-cutter, not the cookie.

@akutz you seem frustrated at conflicting(?) guidance, but I don't know where the conflict is coming from.  As far as I can tell EVERY "SomethingClass" in kubernetes core is non-namespaced, so I don't see the confusion.

I don't know in great detail how or why your management clusters are comingling control plane and data plane operations, but that seems to be at the root of the problem.

A "SomethingClass" is something defined for use by this cluster.  At least, that is the precedent.  Now one could argue that all non-namespaced resources were a mistake, but there are real complexity tradeoffs in the other direction, too.  We've tried to walk a line where things that are most commonly done by admins live in admin space, and things that are most commonly done by app-owners live in app space.  We've not always nailed it, but overall it seems to hold up.

One thing that we discussed a long time ago, but ultimately back-burnered and never came back to is parameterization of classes - the idea that my "class" could declare params and my "claim" could pass those in.  I'm not sure that helps here, but maybe that gets to the distinction between cookie-cutters and cookies, and the surprisingly large cardinality you see?

thockin commented on docs-src/design-decisions.md:
```diff
@@ -0,0 +1,44 @@
+namespace-scoped resource. 
+
+### Advantages of a Namespace-Scoped GatewayClass
+* Less potential for naming conflicts among different implementations.
```
I see this as a non-issue.  Classes should (ideally) be named qualitatively and abstractly, and are offered by a single "authority".

robscott replied:
That's a fair point, I've removed this.

thockin commented on docs-src/design-decisions.md:
```diff
@@ -0,0 +1,44 @@
+
+### Advantages of a Namespace-Scoped GatewayClass
+* Less potential for naming conflicts among different implementations.
+* Enables using Service API on a management cluster designed to provision and
```
comment got lost, recreating from memory.

I object to "management cluster" as a term of art without a definition.  I don't think there's any established meaning or semantic expectations for it.

akutz replied:
https://cluster-api.sigs.k8s.io/reference/glossary.html#management-cluster

robscott replied:
Good call. I've added a link to the management cluster definition above.

thockin commented on docs-src/design-decisions.md:
```diff
@@ -0,0 +1,44 @@
+* Less potential for naming conflicts among different implementations.
+* Enables using Service API on a management cluster designed to provision and
+  manage other clusters and infrastructure needed by other clusters.
+* Easier to provide simple and secure defaults (only allow Gateways and Routes
```
I think "simple and secure" is very biased wording.  It would allow a different form of access control that what non-namespaced classes allow, but "different" != "simpler" or "more secure"

robscott replied:
Good feedback, I've updated this to:
>Enables a more restrictive (and potentially more secure) default for where a
  GatewayClass could be used. This could mean only allowing Gateways and Routes
  in the same namespace by default.

thockin commented on docs-src/design-decisions.md:
```diff
@@ -0,0 +1,44 @@
+  manage other clusters and infrastructure needed by other clusters.
+* Easier to provide simple and secure defaults (only allow Gateways and Routes
+  in the same namespace by default).
+* GatewayClass parameters can live in the same namespace.
```
If you consider "the root" and "each namespace" as "contexts", this statement is moot -- why can't GatewayClass parameters exist in the root context?

robscott replied:
I've removed this point and added a somewhat similar one:

>Namespaces could be used to separate different GatewayClasses, enabling admins
  to more closely restrict who can modify individual GatewayClasses.

Assuming GatewayClass parametersRef is still a local object reference, this would allow admins to easily restrict who could modify certain GatewayClasses _and_ their parameters. With a cluster-scoped class resource, users can either modify all GatewayClasses + parameters or none.

akutz commented:
Hi @thockin,

> The class is the cookie-cutter, not the cookie.

I think it might be more accurate to state:

* The cookie-cutter works for a single batch of cookie dough and
* The class is the cookie-cutter, not the cookie

The cookie dough in this case would be the Amazon ELB service or the GCE analogue. You have a single service that provides a *massive* batch of cookie dough for everyone in the world, and you can use classes, cookie cutters, with that cookie dough. You might have a cookie cutter in the shape of a heart, or a tree. But the point is that the cookie cutters are all leveraging this well established, ever present batch of cookie dough.

Now, imagine for a moment that that batch of cookie dough does not exist. What if you are holding an event with 500 kids, with groups of 10 to a table. That's 50 tables, and as the host your job is to make sure that each table will have:

* Its own batch of cookie dough
* Its own collection of cookie cutters

Not only that, you need to test all those cookie cutters and all those batches of dough before handing them out to those 50 tables.

That's the situation with the class pattern and management clusters. It's fine when you have a *massive* table with a *massive* batch of dough and a few cookie cutters all guests can share. But that's not the case when you lack that single, massive batch of cookie dough.

In the case of `GatewayClass`, the dough in my use case is HAProxy. Every `Cluster` gets its own HAProxy VM. The HAProxy VM **is** the cookie dough. And if I, the host of the event, have to also leverage that HAProxy VM, then I have to implement a `GatewayClass` on the management cluster that points to that HAProxy VM.

The class pattern is only successful when everything that uses it shares the same, backend infrastructure used by one or more related classes. But when the infrastructure itself scales the way we expect `Cluster` resources to scale, the class pattern needs to scale with it, and as a cluster scoped resource, that starts to look very frunklish.

p.s. I did not know what word I wanted to use at the end, so I just made one up that I think works well :)

akutz commented:
Hi all,

I've been told I can share this:

Kubernetes on vSphere 7 implements something called [the _paravirtual_ model](https://docs.vmware.com/en/VMware-vSphere/7.0/vmware-vsphere-with-kubernetes/GUID-D79D0D72-8172-4C97-AB24-F6A039BBA0FF.html). This model is such that:

1. a guest cluster is deployed by the Supervisor (management) cluster
2. it is given a service account that enables its components (cloud provider, CSI driver, etc.) to communicate back to the Supervisor cluster
3. when a target cluster needs a resource typically provided by the cloud provider, CSI, etc., the provider in the guest cluster creates a resource in the Supervisor cluster with the service account from the previous step
4. the supervisor cluster has a controller that watches said resource and reconciles it into a load balancer (if one does not exist), virtual server, a storage volume, etc.
5. the guest cluster can now use the resource

The above model is how the cloud provider in Kubernetes on vSphere 7 works. The cloud provider in guest clusters create a resource on the Supervisor cluster that is reconciled into the virtual server for the service on the guest cluster. 

That means if there is a load balancer **per** `Cluster` resource on the Supervisor cluster, then there also needs to be a `GatewayClass` **per** `Cluster` resource on the Supervisor cluster if the infrastructure for said `GatewayClass` has independent control planes, ex. HAProxy VMs per guest cluster. 

This model means that none of the hundreds, thousands of guest clusters that are provisioned by Project Pacific need to have credentials to talk directly to vSphere, NSX, HAProxy, or any other infrastructure provisioned for the guest clusters. All of the actual infra management and resource management on that infra is handled on the Supervisor (management) cluster on behalf of the guest clusters.

thockin commented:
First, the cookie cutter analogy is getting hilarious, so thanks for that.

> The cookie-cutter works for a single batch of cookie dough

So your user cluster creates a namespaced Gateway in the admin cluster, referencing a class.  I see two approaches for class:
1) reference `class=coke-ha-proxy` and `class=pepsi-ha-proxy` distinctly; or
2) reference `class=ha-proxy` and let the controller decide, based on the original Gateway, whether it is `coke` or `pepsi` that triggered the action.

Does that hold water?

akutz commented:
> First, the cookie cutter analogy is getting hilarious, so thanks for that.
> 
> > The cookie-cutter works for a single batch of cookie dough
> 
> So your user cluster creates a namespaced Gateway in the admin cluster, referencing a class. I see two approaches for class:
> 
> 1. reference `class=coke-ha-proxy` and `class=pepsi-ha-proxy` distinctly; or
> 2. reference `class=ha-proxy` and let the controller decide, based on the original Gateway, whether it is `coke` or `pepsi` that triggered the action.
> 
> Does that hold water?

Where does the information about how to access the distinct HAProxy backends come from then? The Gateway? Isn't this anti-pattern to the way the GatewayClass is positioned to be used?

akutz commented:
Classes related to infrastructure do not exist in a vacuum, and it is difficult to reconcile the notion of a _few_ cluster-scoped class resources with systems involving exponentially greater numbers of the infrastructure that enable the classes. 

There are aways around this:

* Create a field `InfraRef *LocalObjectReference` on a `Gateway` class that can be used to point to the access information for the load balancer to which the `Gateway` belongs
* Make the field `ParametersRef` a list of `*ObjectReference` resources, and define some key that maps from a `Gateway` to select which one of those parameters packages to use with that `Gateway`

But I would argue each of these just handle the same issue in different, less useful ways. When the `GatewayClass` represents a cookie cutter, it does so based on some backend authority from which to cut those cookies. So if you have one `GatewayClass` that you use to access *multiple* HAProxy load balancers, you're also implying that there is some authority shared between those HAProxy load balancers, and there likely won't be. Each one of those HAProxy load balancers has its own config, is its own control plane.

Which is why, if a `GatewayClass` represents some authority, then you need a 1x1 relationship with the `GatewayClass` and that authority, the load balancer platform's control plane. When the latter exists *per* load balancer, so to must a `GatewayClass` exist *per* load balancer.

thockin commented:
> Where does the information about how to access the distinct HAProxy backends come from then? 
> The Gateway? Isn't this anti-pattern to the way the GatewayClass is positioned to be used?

I'm making an unstated assumption, so let me state it.  Within the context of the manager cluster, the namespace that is assigned to the user-cluster might be used from the user cluster's control plane (e.g. cloud-controller-manager) but that use is controlled.  It might be able to make a Gateway, but it can not add a label to its namespace, and you might have other types (e.g. CRDs, secrets) that the user-cluster can not access.

Assuming that assumption is close to true, you can squirrel away information in that namespace that is associated 1:1 with the namespace but is still "safe" from anyone in that user cluster.

So you can extract information about which tenant that namespace is for - labels, annotations, CRDs, secrets etc.

Your multi-tenant Gateway controller (in the admin cluster) can then look up that information and use it when acting on behalf of that tenant.

akutz commented:
This doesn't scale when infra

> > Where does the information about how to access the distinct HAProxy backends come from then?
> > The Gateway? Isn't this anti-pattern to the way the GatewayClass is positioned to be used?
> 
> I'm making an unstated assumption, so let me state it. Within the context of the manager cluster, the namespace that is assigned to the user-cluster might be used from the user cluster's control plane (e.g. cloud-controller-manager) but that use is controlled. It might be able to make a Gateway, but it can not add a label to its namespace, and you might have other types (e.g. CRDs, secrets) that the user-cluster can not access.
> 
> Assuming that assumption is close to true, you can squirrel away information in that namespace that is associated 1:1 with the namespace but is still "safe" from anyone in that user cluster.
> 
> So you can extract information about which tenant that namespace is for - labels, annotations, CRDs, secrets etc.
> 
> Your multi-tenant Gateway controller (in the admin cluster) can then look up that information and use it when acting on behalf of that tenant.

You just defined how I do that in CAPV for the HAProxy load balancer there. Assumed Secret resource names. But in this case there isn't a `Cluster` per namespace. There may be multiple Clusters per namespace. What you're suggesting is similar to CAPV -- use the name of the Cluster to assume some implied resource to access the LB.

I ask then, how does the controller for the GatewayClass know what to do? What you proposed is anti-pattern for how the controller works today (we have an end-to-end solution with HAProxy). 

```yaml
apiVersion: networking.x-k8s.io/v1alpha1
kind: GatewayClass
metadata:
  name: haproxy-lb
spec:
  controller: lbapi.run.tanzu.vmware.com/haproxy
  parameters:
    group: v1
    resource: Secret
    name: haproxy-lb-config
```

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: haproxy-lb-config
  namespace: default
stringData:
  value: |
    debug: true
    servers:
    - https://172.17.0.2:5556/v2
    serverName: dataplane-api
    certificateAuthorityData: "$({ base64 -w0 2>/dev/null || base64; } <hack/images/haproxy/example/ca.crt)"
    clientCertificateData: "$({ base64 -w0 2>/dev/null || base64; } <hack/images/haproxy/example/client.crt)"
    clientKeyData: "$({ base64 -w0 2>/dev/null || base64; } <hack/images/haproxy/example/client.key)"
```

This approach is consistent with the Service API, it just doesn't scale with cluster-scoped resources. I want to make sure that the implementation is consistent with the API. If there should be something specific to the LB then it should be modeled, even as an optional bit, so it can be accounted for in the API's contract.

thockin commented:
> So if you have one GatewayClass that you use to access multiple HAProxy load balancers, you're also implying that there is some authority shared between those HAProxy load balancers, and there likely won't be.

They exist in the same admin cluster, ostensibly that cluster represents the authority.

In my mind this is no different that when your AWS account talks to the AWS API to create an ELB.  I'd bet 1 hundred bucks that AWSs implementation knows "Hey that "create ELB" came from the 'akutz' account, let me look at side-band data we have about that account to make a decision pertaining to the instance of the ELB he asked us to create". 

If the admin-cluster is the cloud provider for the user-cluster, just change the words.  "Hey that "create Gateway" came from the 'akutz' namespace, let me look at side-band data we have about that namespace to make a decision pertaining to the instance of the Gateway he asked us to create".

jpeach commented:
> That means if there is a load balancer **per** `Cluster` resource on the Supervisor cluster, then there also needs to be a `GatewayClass` **per** `Cluster` resource on the Supervisor cluster if the infrastructure for said `GatewayClass` has independent control planes, ex. HAProxy VMs per guest cluster.

I don't follow this at all. The purpose of the GatewayClass is for the guest cluster to be able to instantiate Gateways. So the GatewayClass should be a cluster resource inside the guest cluster, and there only needs to be one of them to be able to barge resources out of vSphere. You can track the guest cluster's load balancing resources in the supervisor cluster however you want. There is no requirement to use GatewayClass for this (or benefit that I can see).

In your example, the guest cluster creates a Gateway. That Gateway references the per-cluster vSphere GatewayClass. That GatewayClass is backed by a controller that has been provisioned with the service account. The controller has some understanding of the mapping between Gateways and vSphere load balancers. If it needs a vSphere load balancer, it creates the relevant supervisor cluster resource (which would not be a GatewayClass, but an infrastructure specific type). Reconciliation ensues. Eventually, the guest cluster controller has enough resources that it can instantiate the Gateway.

akutz commented:
> They exist in the same admin cluster, ostensibly that cluster represents the authority.

The authority is the HAProxy load balancer's configurations file which lives on the HAProxy VM.

jpeach commented:
> > So if you have one GatewayClass that you use to access multiple HAProxy load balancers, you're also implying that there is some authority shared between those HAProxy load balancers, and there likely won't be.
> 
> They exist in the same admin cluster, ostensibly that cluster represents the authority.
> 
> In my mind this is no different that when your AWS account talks to the AWS API to create an ELB. I'd bet 1 hundred bucks that AWSs implementation knows "Hey that "create ELB" came from the 'akutz' account, let me look at side-band data we have about that account to make a decision pertaining to the instance of the ELB he asked us to create".
> 
> If the admin-cluster is the cloud provider for the user-cluster, just change the words. "Hey that "create Gateway" came from the 'akutz' namespace, let me look at side-band data we have about that namespace to make a decision pertaining to the instance of the Gateway he asked us to create".

This is similar to my [original suggestion](https://github.com/kubernetes-sigs/service-apis/issues/136#issuecomment-612570488) but phrased better :)

akutz commented:
> I don't follow this at all. The purpose of the GatewayClass is for the guest cluster to be able to instantiate Gateways. So the GatewayClass should be a cluster resource inside the guest cluster, and there only needs to be one of them to be able to barge resources out of vSphere. You can track the guest cluster's load balancing resources in the supervisor cluster however you want. There is no requirement to use GatewayClass for this (or benefit that I can see).

This is not the way it works at all. Please read the above example about the paravirtual model.

jpeach commented:
> > I don't follow this at all. The purpose of the GatewayClass is for the guest cluster to be able to instantiate Gateways. So the GatewayClass should be a cluster resource inside the guest cluster, and there only needs to be one of them to be able to barge resources out of vSphere. You can track the guest cluster's load balancing resources in the supervisor cluster however you want. There is no requirement to use GatewayClass for this (or benefit that I can see).
> 
> This is not the way it works at all. Please read the above example about the paravirtual model.

Can you explain where this is wrong?

akutz commented:
> So the GatewayClass should be a cluster resource inside the guest cluster, and there only needs to be one of them to be able to barge resources out of vSphere.

> Can you explain where this is wrong?

Read the paravirtual use case I stated above. The first thing you said is a complete 180 from what happens in that use case.

thockin commented:
Let's keep it constructive, please.

> You just defined how I do that in CAPV for the HAProxy load balancer -- implied Secret resource names.

Great!  The chances of us BOTh being nuts is much lower.

> But in this case there isn't a Cluster per namespace. There may be multiple Clusters per namespace. What you're suggesting is similar to CAPV -- use the name of the Cluster to assume some implied resource to access the LB.

OK.  Options:

1) Don't do that (doctor, it hurts when I ...)
2) Pass a cluster-id in as an annotation or label
3) Parameterize classes (heretofore not required, but has been considered)

> I ask then, how does the controller for the GatewayClass know what to do? What you proposed is anti-pattern for how the controller works today (we have an end-to-end solution with HAProxy).

What you are asking for is how to subdivide a single namespace.  That's kind of
what labels are FOR.

akutz commented:
> What you are asking for is how to subdivide a single namespace. That's kind of
what labels are FOR.

What I'm asking is to sub-divide a cluster, which is what namespaces are for :)

akutz commented:
To put it another way, the `GatewayClass` is the hub through which all things Service API reconciliation flows. What's the value in that resource existing at the cluster scope if the infrastructure tied to it isn't available to the entire cluster?

thockin commented:
> To put it another way, the GatewayClass is the hub through which all things Service API reconciliation flows. What's the value in that resource existing at the cluster scope if the infrastructure tied to it isn't available to the entire cluster?

The infrastructure IS available.  Presumably any cluster of any tenant, modulo quota and so forth, could request an instance of a Gateway of any class.

To return to the cookies analogy: 

I'm saying that a good cookie-cutter can cut a cookie from any flavor of dough, you just need to provide the dough.  You're saying that the dough is intrinsic to the cutter.

The GatewayClass is the cutter, the controller is the hand that uses the cutter, and the dough comes from context in or around the Gateway that started it all.

akutz commented:
> The infrastructure IS available. Presumably any cluster of any tenant, modulo quota and so forth, could request an instance of a Gateway of any class.

Isn't that because thus far infrastructure in general is ever present, available to the entire cluster, and can be ignored as something necessary to enable the class?

akutz commented:
I'm exhausted. It is unclear whether all the failures at communication have happened on my end, on the other end, or, as I suspect, somewhere in between.

I am trying to argue on behalf of a design not my own, but one that is I believe has merit, has much broader implications in a post-Cluster API world, and one that I strongly believe believe this API should support. For those reading **this** comment, https://github.com/kubernetes-sigs/service-apis/pull/156#issuecomment-617413324 is where I outline the use case as clearly as I can state it. 

Luckily the team who did design the paravirtual model just told me they are willing to accept the cluster-scoped resources non the Supervisor cluster. I think primarily just to shut me up. Y'all should send them a cookie bouquet or something :)

@robscott, many thanks to you, @thockin and @youngnick for your help.

akutz commented:
/lgtm

thockin commented:
> I'm exhausted. It is unclear whether the failures at communication have happened on my end, or, as I suspect, somewhere in between.

I don't like "winning" by attrition.  That's not how we get the best ideas.

I'm not sure what the breakdown is either - I do feel like I understand your model, I just don't agree with it.

I had not seen your comment update here, so let me try one last time to see if there's something there that I missed:

> 1. a guest cluster is deployed by the Supervisor (management) cluster
> 2. it is given a service account that enables its components (cloud provider, CSI driver, etc.) to communicate back to the Supervisor cluster

Unstated here, but stated elsewhere: In the supervisor cluster, one namespace
is one "tenant" which may have any number of user-clusters in it.  Right?

> 3. when a target cluster needs a resource typically provided by the cloud provider, CSI, etc., the provider in the guest cluster creates a resource in the Supervisor cluster with the service account from the previous step
> 4. the supervisor cluster has a controller that watches said resource and reconciles it into a load balancer (if one does not exist), virtual server, a storage volume, etc.
> 5. the guest cluster can now use the resource

So TL;DR: you have made a new cloud-provider that is implemented as another
kubernetes cluster.  Presumably each user-cluster is mapped to a single
supervisor-namespace, and (as above) each supervisor-namespace maps to N
user-clusters.

> That means if there is a load balancer **per** `Cluster` resource on the Supervisor cluster, then there also needs to be a `GatewayClass` **per** `Cluster` resource on the Supervisor cluster if the infrastructure for said `GatewayClass` has independent control planes, ex. HAProxy VMs per guest cluster. 

This is where you lost me.  I can see how that could work, but I don't see how
it is NECESSARILY the only model.

To create a new HTTP route for a given user-cluster in the correct LB instance
for that cluster you need a tuple of {LB-category, originating-cluster, route},
right?

You can bake the first 2 items into GatewayClass and the last item into
Gateway, which is what I think you are expressing.

Or you can bake the first into GatewayClass, the last 2 into Gateway.  I
proposed elsewhere at least 2 ways to do that.  This fits more closely with how
Class types work today.

There's no loss of information, and while you called it an anti-pattern, I
don't think I agree.

If you look at it from the perspective of a "user" of your new "cloud
provider": I can say `kubectl get gatewayclass` and see all of the available
classes and they MEAN SOMETHING.  A single gateway class, on its own, does not
have enough information to do anything, but when you couple it with the orginal
Gateway, it does.

robscott commented:
Thanks for all the great feedback here, there's a lot to digest. I've updated my PR to try to include the key arguments in favor of namespace-scoped and cluster-scoped resources, though I've likely still missed some, please let me know if that's the case. I've also added a link to the discussion in this PR for anyone wanting the full context around this decision.

One point that I think may have been missed was raised earlier by @aledbf. With Ingress, users could deploy a full stack, including a controller, with only namespace-scoped access. With a cluster-scoped GatewayClass (and IngressClass) resource we're removing that ability for a number of users. The first concept of `class` in Ingress was simply an annotation on an Ingress resource and thus was namespace-scoped by default. Before we finalize any decision here, I want to be sure that we are OK with the impact of this change.

thockin commented:
Re: DIY ingress in a namespace:  At what point does that become more complex than it is worth?

But forget the philosophy - can't you still achieve this, just without the actual "GatewayClass" type being used?

E.g. I can deploy my private controller and tell it to listen for class "xx-tims-private-gateway-xx".  I can create Gateway instances that use "className: xx-tims-private-gateway-xx".  There's not actually a GatewayClass called "xx-tims-private-gateway-xx", but that's OK.  It's a little different, but we don't do referential integrity anyway.

robscott commented:
@thockin That is an interesting way to approach this. Certainly not at all what was intended, but also entirely possible and generally equivalent to the previous Ingress class annotation. Agree that DIY Ingress per namespace, especially DIY Gateway per namespace, is likely more complex than it's worth in most cases.

robscott commented:
/retest

robscott commented:
@thockin @akutz @jpeach @youngnick A few follow up questions to see if we can finalize this PR:
1. Does it still make sense to document this decision? 
2. Is this an approach everyone is OK with? 
3. Is this PR a sufficient summary of the key points behind this decision?

akutz commented:
Hi @robscott,

My position is well-documented and discussed at this point. I understand the counter-argument to GWC being namespaced, I just happen to disagree with it. It does not hurt anyone, and helps others. To me that's a win-win. I think it is worth codifying, because others in the future may come to this project with the same concerns.

robscott commented:
Completely understand @akutz - is there anything I've missed in the contents of this PR as far as the benefits of either approach? Want to make sure this does an adequate job at covering the main points in this discussion.

akutz commented:
> is there anything I've missed in the contents of this PR as far as the benefits of either approach?

Nope. Lgtm.

robscott commented:
I think there's a better way to document this, likely an addition to concepts doc and/or references to the discussion here. Closing this PR.

/close
